internal/httputil/certpoolwatcher.go

```diff
@@ -0,0 +1,108 @@
+package httputil
+
+import (
+	"crypto/x509"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-logr/logr"
+)
+
+type CertPoolWatcher struct {
+	generation int
+	dir        string
+	mx         sync.RWMutex
+	pool       *x509.CertPool
+	log        logr.Logger
+	watcher    *fsnotify.Watcher
+	done       chan bool
+}
+
+// Returns the current CertPool and the generation number
+func (cpw *CertPoolWatcher) Get() (*x509.CertPool, int, error) {
+	cpw.mx.RLock()
+	defer cpw.mx.RUnlock()
+	if cpw.pool == nil {
+		return nil, 0, fmt.Errorf("no certificate pool available")
+	}
+	return cpw.pool.Clone(), cpw.generation, nil
+}
+
+func (cpw *CertPoolWatcher) Done() {
+	cpw.done <- true
+}
+
+func NewCertPoolWatcher(caDir string, log logr.Logger) (*CertPoolWatcher, error) {
+	pool, err := NewCertPool(caDir, log)
+	if err != nil {
+		return nil, err
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err = watcher.Add(caDir); err != nil {
+		return nil, err
+	}
+
+	cpw := &CertPoolWatcher{
+		generation: 1,
+		dir:        caDir,
+		pool:       pool,
+		log:        log,
+		watcher:    watcher,
+		done:       make(chan bool),
+	}
+	go func() {
+		for {
+			select {
+			case <-watcher.Events:
```

Comment:
Could there be spurious events in this dir that are going to cause us to do work when we don't really need to? Just wondering how clean cert manager, and anything else with access, is in terms of usage of this dir? Maybe we should filter it down further or just watch .crt files, etc.

Reply:
That's what this function is for... it's used to "drain" the event cue to avoid spurious events.
It's really just kubernetes that has write access, it updates the certs based on when a Secret is updated. The directory is read-only within the pod.

Reply:
sounds like it'll be okay then